opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py

```diff
@@ -78,6 +83,126 @@ def shutdown(self) -> None:
         self.span_exporter.shutdown()
 
 
+class BatchExportSpanProcessor(SpanProcessor):
+    """Batch span processor implementation.
+
+    BatchExportSpanProcessor is an implementation of `SpanProcessor` that
+    batches ended spans and pushes them to the configured `SpanExporter`.
+    """
+
+    def __init__(
+        self,
+        span_exporter: SpanExporter,
+        max_queue_size: int = 2048,
+        schedule_delay_millis: float = 5000,
+        max_export_batch_size: int = 512,
+    ):
+        if max_queue_size <= 0:
+            raise ValueError("max_queue_size must be a positive integer.")
+
+        if schedule_delay_millis <= 0:
+            raise ValueError("schedule_delay_millis must be positive.")
+
+        if max_export_batch_size <= 0:
+            raise ValueError(
+                "max_export_batch_size must be a positive integer."
+            )
+
+        if max_export_batch_size > max_queue_size:
+            raise ValueError(
+                "max_export_batch_size must be less and equal to max_export_batch_size."
+            )
+
+        self.span_exporter = span_exporter
+        self.queue = collections.deque([], max_queue_size)
+        self.worker_thread = threading.Thread(target=self.worker, daemon=True)
+        self.condition = threading.Condition(threading.Lock())
+        self.schedule_delay_millis = schedule_delay_millis
+        self.max_export_batch_size = max_export_batch_size
+        self.max_queue_size = max_queue_size
+        self.done = False
+        # flag that indicates that spans are being dropped
+        self._spans_dropped = False
+        # used to avoid sending too much notifications to worker
+        self.notified = False
+        self.worker_thread.start()
+
+    def on_start(self, span: Span) -> None:
+        pass
+
+    def on_end(self, span: Span) -> None:
+        if self.done:
+            logging.warning("Already shutdown, dropping span.")
+            return
+        if len(self.queue) == self.max_queue_size:
+            if not self._spans_dropped:
+                logging.warning("Queue is full, likely spans will be dropped.")
```

Comment:
This is a consequence of adding items outside the condition, because the worker might wake up and consume some spans between here and `appendleft` below?

I bet we'll want exact stats on dropped spans in the future, will probably need to change this behavior.

Reply:
I think if we want exact stats we can either:
* Go back to using `queue.Queue` 
* Use a lock and a plain list
* Estimate the number with `n_ended_spans - (n_exported_or_exporting_spans + len(self.queue))` (becomes exact after shutdown, or whenever no spans are currently added or processed)
* Maybe the queue implementation that @reyang suggested, I haven't looked at it though so I don't know if that would solve the problem.

Reply:
@c24t yes you right. There is not a way to actually now if the queue will be full when `appendleft` will be called. I agree with both of you, if we need to support that use case we'll need another approach.